astropy/utils/xml/src/iterparse.c

```diff
@@ -40,13 +40,56 @@
 static Py_ssize_t
 next_power_of_2(Py_ssize_t n)
 {
-    /* Calculate the next-highest power of two */
+    /* Calculate the next-higher power of two that is >= 'n' */
+
+    /* These instructions are intended for uint32_t and originally 
+       from http://www-graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+       Py_ssize_t is the same as the C datatype ssize_t and is 
+       32/64-bits on 32-bit and 64-bit systems respectively. The implementation
+       here accounts for both.
+       
+       Limitations: Since a signed size_t (ssize_t) was required for the underlying CPython implementation,
+       on 32 bit systems, it will not be possible to allocate memory sizes between
+       SSIZE_MAX and SIZE_MAX (i.e, for allocations in the range [2^31, 2^32 - 1] bytes)
+       even though such memory might be available and accessible on the (32-bit) computer. That
+       said, since the underlying CPython implementation *also* uses Py_ssize_t (i.e., ssize_t),
+       it is safe to assume that such memory allocations would probably not be usable anyway. 
+       
+       TLDR: Will work on 64-bit machines but be careful on 32 bit machines when reading in 
+       ~ 2+ GB of memory -- @manodeep 2020-03-27
+    */
+
+    /* If negative numbers are also allowed, then we need this line: 
+    if(n <= 1) {
+        return 1;
+    } 
+    */
+    
     n--;
+
+#if 1
+    /* This section just extends the existing solution to 64 bit systems */
     n |= n >> 1;
     n |= n >> 2;
     n |= n >> 4;
     n |= n >> 8;
     n |= n >> 16;
+    if(sizeof(Py_ssize_t) > 4) {
```

Comment:
More of a question that anything else: this would be optimized away by the compiler, right? Since presumably it is fixed on any given system anyway. Might it be an idea regardless to make it a pre-compile `#if ...` so that that is obvious? With that, I think I'd stick with this simpler solution instead of the loop.

Reply:
Yeah - I think that `if` condition will (should?) get eliminated by all reasonable compilers under moderate optimisation. 

However, rewriting that `if` condition as a compile-time condition is a deceptively complicated task! `sizeof` calls are evaluated by the compiler and not the pre-processor. I looked into other possible ways of detecting 64-bit memory-addressing (you can have 64-bit addressing in a 32-bit system but I don't know what value `Py_ssize_t` takes in that case), but all the ones I found were not fully portable or restricted to specific compilers (`gcc` macros), common OS'es etc. 

The one solution I can think of is using:

```C
// at the top of the file
#ifndef PY_SSIZE_T_MAX
#define PY_SSIZE_T_MAX (SSIZE_MAX)
#endif

//Check if maximum addressable is > 2^31-1 (==0x7FFFFFFF)
#if PY_SSIZE_T_MAX > 0x7FFFFFFF
    n |= n >> 32;
#endif
```

But this solution depends on `SSIZE_MAX` being defined. While `SSIZE_MAX` is defined on all POSIX systems, I am unsure what happens on Windows.

Basically, I am not confident in the pre-processor solution and it might break the build. I ma happy to give this solution a shot, and revert the fix if the CI chokes.

Reply:
That's curious.  I just checked numpy, and it uses a `__LP64__`:
https://github.com/numpy/numpy/blob/f0a74b2d4e50a1b4f9d9189c6b2e31b920913a8b/numpy/core/include/numpy/numpyconfig.h#L15-L22

That seems some kind of standard; see https://stackoverflow.com/questions/6721037/where-is-lp64-defined-for-default-builds-of-c-applications-on-osx-10-6

Reply:
That's specifically for OSX. You will need similar definitions on all supported OS'es + compilers. 

For most systems, the presence of any of the macros `__x86_64__`, `__LP64__`, `__amd64__`, `__IA64__` indicates that 64-bit compilation is in effect. On Windows, `_M_X64`, or `_WIN64` indicates 64-bit compilation. On ARM64 systems, the GNU C standard defines `__aarch64__` (think that requires compiling with `-std=gnu*`). Unsure what happens on Cray, Solaris, other "unusual" OS'es. [Source for pre-defined macros on various systems](https://sourceforge.net/p/predef/wiki/Architectures/).

@mhvk I am happy to code that in using `__LP64__` if you think that will work within the astropy-supported OS'es. However, in trying to figure out how to properly code this at the pre-processor stage, I am somewhat certain that the current implementation with the `if` condition is the *most* robust.

Reply:
Duh, I should really have looked more carefully. The numpy macro undefines and then redefines it for apple, but I cannot even find where it is defined in the first place. In princple, I guess we could import the numpy macros and just use theirs. But I am starting to agree that we might as well let the compiler optimize it away! Or should one just unconditionally run the extra instruction? If the compiler is smart enough it will realize that one is or'ing with 0 and optimize that away as well...

Anyway, I feel I'm causing you far more work than it's worth! Just go ahead and do what you feel is best, and let's get this in!

Reply:
It was a fun exercise for the pedant in me :D. Really did not expect implementing the `if` condition as a (portable) pre-processor condition to be this difficult!

Bit-shifting in excess of the bit-width of a variable (that would occur on 32-bit systems if we always execute the last right-shift by 32 bits) will be undefined behaviour and allows the compiler complete freedom to replace/remove/do anything the compiler wants. Think it's best to stick to the `if` condition.

Reply:
OK, so let's go with the explicit solution. Could you remove the loop part?  I think it might also be good to rewrite the comment a bit: the solution is now good for either 32 or 64 bit, but on 32 bit cannot use the highest one.